### PR TITLE
feat: roxctl --config-file flag for instance management

### PR DESCRIPTION
**Pull request by randyttruong**
### Description

When using `roxctl` to communicate with various Stackrox instances, users must manually adjust values for 
- `ROX_ENDPOINT` (`-e` / `--endpoint`) 
- `ROX_CA_CERT_FILE` (`--ca`) 
- `ROX_API_TOKEN` (`--token-file`)

as per the issue raised in #12157.

The proposed solution is to create a `--config-file` flag that consumes a YAML file and provides values to the previously-mentioned flags.

In short, my solution creates a new flag file `/roxctl/common/flags/config_file.go` that defines a new persistent flag `--config-file` as well as a PreRun function for the root `roxctl` command that loads the file by replacing existing flag values with those in the YAML file. 

As value prioritization, my current implementation has flags > configuration > environment variables. As such, I have implemented a new method `flagOrConfigurationValue` with this prioritization in mind, while still keeping the original `flagOrSettingValue` function as to maintain the original functionality for all non-configuration-editable flags. 

### Checklist 
- [X] Create a new environment variable for the config file path in `/pkg/env/roxctl.go` 
- [X] Create a new flag file `/roxctl/common/flags/config_file.go`
- [ ] Create new kubeconfig like struct to represent configuration file 
     - [X] Prototype for single instance, single credential, and single context 
     - [ ] Multi-instance, multi-credential, and multi-context  
- [ ] Create YAML parsing function and value loader that are preemptively run with `roxctl` command
     - [X]  Prototype for single instance, single credential, and single context 
     - [ ] Custom unmarshaling function for ingesting arrays of instances/authinfo/contexts as maps 
- [X] Create precedence functions 
     - [X] `flagOrConfigurationValue` for inline value priority in `/roxctl/common/flags/env_flag.go` 
     - [X]`flagOrConfigurationValueWithFilepathOption` for precedence between inline and filepath options 
     - [X] `booleanFlagOrConfigurationValue` for precedence between flag/env/configuration boolean options 

### Testing 
- [ ] Configuration file reading + variable loading (`roxctl/common/flags/config_file_test.go`)
- [x] Value prioritization tests (`roxctl/common/flags/env_flag_test.go`)
    - [x] flags > config
    - [ ] env > config 
    - [ ] flags > env > config 
    - [ ] inline > filepath 
    - [ ] boolean flags -> boolean env -> boolean config

### CODE REVIEW

**Review by rhybrillou — COMMENTED** [review]
Thank you for your contribution. I like the idea behind this change.
Before I dive deeper into the code, could you please fix the issues reported by the CI style targets (I marked them in the impacted files) ?

File: roxctl/common/flags/env_flag_test.go [hunk 1]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+	assert.False(t, *endpointChanged)
+
+	err := cmd.PersistentFlags().Set("config-file", testFile1)
+	assert.NoError(t, err)
+	assert.Equal(t, testFile1, ConfigurationFileName())
+
+	err = cmd.PersistentPreRunE(cmd, []string{})
+	assert.True(t, ConfigurationFileChanged())
+
+	// Execute the command to trigger PersistentPreRunE
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:61]: What is the added value of this comment ? PersistentPreRunE is called 2 lines before.

File: roxctl/common/flags/env_flag_test.go [hunk 2]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+
+	cmd.PersistentFlags().Set("ca", "some-other-other-test-value")
+	assert.Equal(t, "some-other-other-test-value", caCertFile)
+
+}
+
+func TestFlagOrConfigurationValueEnv(t *testing.T) {
+	var (
+		testFile1 = "./testdata/test_instance1.yaml"
+	)
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:84]:
```suggestion
	testFile1 := "./testdata/test_instance1.yaml"
```

File: roxctl/common/flags/env_flag_test.go [hunk 3]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+	t.Setenv("ROX_CA_CERT_FILE", "some-other-other-test-env-value")
+	assert.Equal(t, "./deploy/cert", caCertFile)
+
+}
+
+// TODO: Add flag > config > env setting tests
+func TestFlagOrConfigurationValueFlagsAndEnv(t *testing.T) {
+	var (
+		testFile1 = "./testdata/test_instance1.yaml"
+	)
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:128]:
```suggestion
	testFile1 := "./testdata/test_instance1.yaml"
```

File: roxctl/common/flags/env_flag_test.go [hunk 4]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+	}
+
+	AddAPITokenFile(cmd)
+	AddConnectionFlags(cmd)
+	AddConfigurationFile(cmd)
+
+	assert.False(t, ConfigurationFileChanged())
+	assert.False(t, *caCertFileSet)
+	assert.Empty(t, *apiTokenFileChanged)
+	assert.False(t, *endpointChanged)
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:141]: Could all this be extracted to a helper function ? This seems to be re-used across the test functions.

File: roxctl/common/flags/env_flag_test.go [hunk 5]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+
+	t.Setenv("ROX_ENDPOINT", "some-test-env-value")
+	assert.Equal(t, "localhost:8000", endpoint)
+
+	t.Setenv("ROX_CA_CERT_FILE", "some-other-other-test-env-value")
+	assert.Equal(t, "./deploy/cert", caCertFile)
+
+}
+
+// TODO: Add flag > config > env setting tests
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:124]: This seems to be done.

File: roxctl/common/flags/env_flag_test.go
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:1]: Ideally, the naming of the test functions should reflect what they are testing.
From what I read, the added test functions do test the precedence of the values between data sources, which is good. Could you make sure either the test function name, or a comment at the start of the test make clear which precedence rule is being tested ?

File: roxctl/common/flags/env_flag_test.go [hunk 6]
```diff
@@ -30,6 +31,145 @@ func TestFlagOrSettingValue(t *testing.T) {
...
+
+	// 3. Validate that PersistentPreRunE (LoadConfig) ran successfully
+	assert.NotNil(t, config, "Config should be initialized after LoadConfig runs")
+
+	// 4. Change environment values.
+	t.Setenv("ROX_API_TOKEN_FILE", "some-other-test-env-value")
+	t.Setenv("ROX_ENDPOINT", "some-test-env-value")
+	t.Setenv("ROX_CA_CERT_FILE", "some-other-other-test-env-value")
+
+	// 5. Change flag values. Flag values should be returned, irrespective of configuration or environment-defined values.
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:161]: I think it would be good to test the precedence of config file settings over env before setting the flags, then to test the precedence of flags over config file settings.

**Review by rhybrillou — COMMENTED** [review]
Very good work overall.

Could you please ensure that the comments in the PR are adjusted or removed to make the code look merge-ready ?
There will likely be another review pass on Tuesday. If so, points like the structure of the config file and the precedence rules will be re-discussed then.

File: roxctl/common/flags/config_file.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import (
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"os"
+
+	"github.com/stackrox/rox/pkg/pointers"
+)
```
Comment (rhybrillou) [inline roxctl/common/flags/config_file.go:12]: Please fix the order of imports and check that all of them are used and none are missing.

```suggestion
import (
	"os"

	"github.com/ghodss/yaml"
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"github.com/stackrox/rox/pkg/env"
	"github.com/stackrox/rox/pkg/logging"
	"github.com/stackrox/rox/pkg/pointers"
)
```

File: roxctl/common/flags/config_file.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/spf13/cobra"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/pointers"
+)
+
+type Instance struct {
+	ProfileName       string `json:"name"`
+	CaCertificatePath string `json:"caCertPath"`
+	ApiTokenFilePath  string `json:"apiTokenPath"`
```
Comment (parametalol) [inline roxctl/common/flags/config_file.go:17]: What about allowing inline CA cert and token? For the latter, I'd suggest an additional check for the 600 file permissions and ownership.
```suggestion
	CaCertificatePath string `json:"caCertPath,omitempty"`
	CaCertificate string `json:"caCert,omitempty"`
	ApiTokenFilePath  string `json:"apiTokenPath,omitempty"`
	ApiToken  string `json:"apiToken,omitempty"`
```

File: roxctl/common/flags/config_file.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	Endpoint          string `json:"endpoint"`
+}
```
Comment (parametalol) [inline roxctl/common/flags/config_file.go:19]: We support other connection parameters, such as:
- `--insecure-skip-tls-verify` and `--insecure`
- `--direct-grpc`
- `--force-http1`
- etc.

I think the config file should support all the same options as the CLI parameters.

Comment (parametalol) [reply roxctl/common/flags/config_file.go:19]: Ideally, the structure would be runtime or somehow automatically generated so that it is in sync with the CLI connection parameters.

File: roxctl/common/flags/config_file.go (hunk 8 above)
Comment (parametalol) [inline roxctl/common/flags/config_file.go:17]: One may need to use different tokens when connecting to the same instance. I suggest adding another level of indirection by extracting credentials (token and password) to a separate structure. Something like:
```go
type Credential struct {
    Name string
    Password string
    ApiToken string
}

type Context struct {
    Name string `json:"name"`
    Context struct{
        InstanceName string `json:"instance"`
        CredentialName string `json:"credential,omitempty"`
    } `json:"context"`
}
```

Please see the kubeconfig structure for inspiration.

File: roxctl/common/flags/config_file.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/pointers"
+)
+
+type Instance struct {
+	ProfileName       string `json:"name"`
```
Comment (parametalol) [inline roxctl/common/flags/config_file.go:15]:
```suggestion
	InstanceName       string `json:"name"`
```

File: roxctl/common/flags/config_file.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	CaCertificatePath string `json:"caCertPath"`
+	ApiTokenFilePath  string `json:"apiTokenPath"`
+	Endpoint          string `json:"endpoint"`
+}
+
+type instanceConfig struct {
+	Instance Instance `yaml:"instance"`
```
Comment (parametalol) [inline roxctl/common/flags/config_file.go:22]: May we add a version string as a top level field?

File: roxctl/common/flags/config_file.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+		return nil
+	}
+
+	instance, err := readConfig(configFile)
+
+	if err != nil || instance == nil {
+		config = nil
+		log.Errorf("Error reading instance config file: %v", err)
+		return err
+	}
```
Comment (rhybrillou) [inline roxctl/common/flags/config_file.go:117]: The currently implemented behaviour in case the file does not exist is to return an zero config object (no endpoint, no CA file path, no token file path).
What is then the result of the precedence rules in that case ? Would command line flags be considered but not environment variables ?
I think Unit tests could help describe/document the behaviour here.

File: roxctl/common/flags/env_flag_test.go [hunk 13]
```diff
@@ -1,6 +1,7 @@
 package flags
 
 import (
+	// "fmt"
```
Comment (rhybrillou) [inline roxctl/common/flags/env_flag_test.go:4]: Please remember to remove this commented import before merging.

**Review by rhybrillou — COMMENTED** [review]
@randyttruong Once again, thank you for this PR. The description and content are very good and help a lot understanding the goal of the change.

We, maintainers of the stackrox codebase, would like to see a configuration file structure more similar to what kubeconfig currently offers. You can implement the minimal set of features you need, and we'd then extend things later on.

It would be great to see more tests that describe the behaviour of the code on corner cases like empty config file name on the command line or files that do not provide all the config parameters.

File: roxctl/common/flags/config_file.go (hunk 8 above)
Comment (randyttruong) [reply roxctl/common/flags/config_file.go:17]: Gotcha, sounds good. I've been eyeing the `kubeconfig` source code for a bit, and I was debating how closely I should follow it, but all of this makes sense.

I just finished exams, so I'll be working more on this throughout this week!

For now-- should is it essential for me to start adding the other connection parameters for this initial version, as per @parametalol? I was originally planning on working on just getting a baseline merged and extend in future PRs

Comment (parametalol) [reply roxctl/common/flags/config_file.go:17]: Please stick to the minimal set of options. We'd like to do some refactoring to simplify the priority layering of the configuration arguments. So that there is a single source of default values (preferably as an instance of the Instance object), and then env vars and CLI flags to reference them.

Comment (randyttruong) [reply roxctl/common/flags/config_file.go]: Sounds good, does the team agree that the precedence order should be flag > env > config > default?

**Comment by randyttruong** [comment]
@rhybrillou In the spirit of keeping the implementation close to that of kubeconfig, I noticed in the source code that they use custom utility functions for deserializing YAML files into `map[string]*ConfigField`. I plan on replicating this functionality by creating new utility functions under `pkg`. Is this fine?

File: roxctl/common/flags/config_file.go (hunk 8 above)
Comment (parametalol) [reply roxctl/common/flags/config_file.go]: Yes.

From my quick test, the YAML library allows the object, passed to the `Unmarshal()` function, to have predefined values, which won't be overridden by empty input (assuming `omitempty`).
Please consider testing and using this functionality for defining the single source of default configuration values.

**Comment by randyttruong** [comment]
Hey @rhybrillou @parametalol, I added a new prototypical kubeconfig-like struct in `config_file.go`. How does the format look? 

The struct right now only possesses the ability to add a single map of an instance, credentials, as well as a context, as I am toying around implementation ideas for arrays of instances/credentials/contexts. 

As of right now, kubeconfig uses a custom unmarshal function that stores arrays as maps, using a `name` key within each array element as the key for storing it within a map, so I'm thinking of similar workarounds.

File: roxctl/common/flags/config_file.go [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+		"",
+		"Utilize instance-specific metadata defined within a configuration file. "+
+			"Alternatively, set the path via the ROX_CONFIG_FILE environment variable")
+	configFileSet = &c.PersistentFlags().Lookup("config-file").Changed
+}
+
+// ConfigurationFile returns the currently specified configuration file name.
+func ConfigurationFileName() string {
+	return flagOrSettingValue(configFile, ConfigurationFileChanged(), env.ConfigFileEnv)
+}
```
[the remark this answers is not in the document]
Comment (randyttruong) [reply roxctl/common/flags/config_file.go:115]: Gotcha, this is definitely a less messy solution. It looks like there isn't a way to implementing looping over the struct fields without either abandoning structs or by using the `reflect` package, so I'll get to working on a hardcoded solution for now. 

Thanks for the guidance!

**Comment by porridge** [comment]
@randyttruong any updates from your side? Do you intend to continue working on this change?

**Comment by randyttruong** [comment]
@porridge Hi, yes I do intend to still work on this. No updates on my side since I've been in the middle of exams, but I expect to get back to working on it this week

**Comment by porridge** [comment]
@randyttruong this was a long week 😉 Any updates from your side?

**Comment by porridge** [comment]
Please do reopen once you have time to refresh this PR.
